### PR TITLE
RelatedCollectionLinkNormalizer performance improved

### PR DESCRIPTION
**Pull request by pmattmann**
Function `getRelatedCollectionHref` is called several times for the same relation 
- try-catch removed 
- cache per relation if no URI can be calculated for it anyway (instead of trying again and again)

## Old:
- all activities
image
- activities of one camp
image

## New:
- all activities
image
- activities of one camp
image

## New2:
- all activities
image
- activities of one camp
image

#6461

### CODE REVIEW

File: api/src/Serializer/Normalizer/RelatedCollectionLinkNormalizer.php [hunk 1]
```diff
@@ -110,12 +113,15 @@ public function normalize($data, $format = null, array $context = []): null|arra
...
                 continue;
             }
-
-            try {
-                $normalized_data['_links'][$rel] = ['href' => $this->getRelatedCollectionHref($data, $rel, $context)];
-            } catch (UnsupportedRelationException $e) {
-                // The relation is not supported, or there is no matching filter defined on the related entity
-                continue;
+            // Only consider relations with non-null value (object or collection)
+            // (does not yet work, some properties are private and cannot be read out)
```
Comment (carlobeltrame) [inline api/src/Serializer/Normalizer/RelatedCollectionLinkNormalizer.php]: Remove nonfunctional code before merging

**Review by carlobeltrame — APPROVED** [review]
Is the array return really faster than the exception? If so, then *facepalm* PHP...
But did you test whether the performance is also improved similarly with only the caching part?

**Comment by pmattmann** [comment]
> Is the array return really faster than the exception? If so, then _facepalm_ PHP... But did you test whether the performance is also improved similarly with only the caching part?

As far as I know, it is faster in most languages if you don't need exceptions for the control flow. 
You can find a few measurements for PHP here: 
https://php.watch/articles/php-exception-performance#results

File: api/src/Serializer/Normalizer/RelatedCollectionLinkNormalizer.php (hunk 1 above)
Comment (pmattmann) [reply api/src/Serializer/Normalizer/RelatedCollectionLinkNormalizer.php]: Agreed - if I can't get the idea to work, I'll remove the code again.

Comment (pmattmann) [reply api/src/Serializer/Normalizer/RelatedCollectionLinkNormalizer.php]: solved

File: api/src/Serializer/Normalizer/RelatedCollectionLinkNormalizer.php [hunk 2]
```diff
@@ -81,6 +82,8 @@ class RelatedCollectionLinkNormalizer implements NormalizerInterface, Serializer
     use PropertyHelperTrait;
     use ClassInfoTrait;
 
+    private $exactSearchFilterExistsCache = [];
```
Comment (BacLuc) [inline api/src/Serializer/Normalizer/RelatedCollectionLinkNormalizer.php]: I would like a type annotation here that we know whats in this cache.

File: api/src/Serializer/Normalizer/RelatedCollectionLinkNormalizer.php [hunk 3]
```diff
@@ -111,12 +114,17 @@ public function normalize($data, $format = null, array $context = []): null|arra
...
-            } catch (UnsupportedRelationException $e) {
-                // The relation is not supported, or there is no matching filter defined on the related entity
+            // If relation is a public property, this property can be checked to be a non-null value
+            $values = get_object_vars($data);
+            if (array_key_exists($rel, $values) && null == $values[$rel]) {
+                // target-value is NULL
                 continue;
             }
+
+            list($ok, $href) = $this->getRelatedCollectionHref($data, $rel, $context);
```
Comment (BacLuc) [inline api/src/Serializer/Normalizer/RelatedCollectionLinkNormalizer.php]: oh, we are programming go now

File: api/src/Serializer/Normalizer/RelatedCollectionLinkNormalizer.php [hunk 4]
```diff
@@ -111,12 +114,17 @@ public function normalize($data, $format = null, array $context = []): null|arra
...
+            if ($ok) {
```
Comment (BacLuc) [inline api/src/Serializer/Normalizer/RelatedCollectionLinkNormalizer.php]: a go programmer would expect the error handling first.
here it looks like you forgot it.

maybe:
if (!$ok) {
 continue;
}
$normalized_data['_links'][$rel] = ['href' => $href];

i know, its more verbose, but at least explicit like before.

File: api/src/Serializer/Normalizer/RelatedCollectionLinkNormalizer.php [hunk 5]
```diff
@@ -170,21 +178,35 @@ public function getRelatedCollectionHref($object, $rel, array $context = []): st
...
+            return [false, 'The '.$resourceClass.'#'.$rel.' relation does not have both a targetEntity and a mappedBy or inversedBy property'];
         }
 
-        $resourceMetadataCollection = $this->resourceMetadataCollectionFactory->create($relatedResourceClass);
-        $operation = OperationHelper::findOneByType($resourceMetadataCollection, GetCollection::class);
+        $lookupKey = $relatedResourceClass.':'.$relatedFilterName;
+        if (isset($this->exactSearchFilterExistsCache[$lookupKey])) {
+            $result = $this->exactSearchFilterExistsCache[$lookupKey];
+        } else {
+            $result = [null, ''];
```
Comment (BacLuc) [inline api/src/Serializer/Normalizer/RelatedCollectionLinkNormalizer.php]: a union type would be enough.
sadly, both types are string...but never both parts of the "tuple" are set.
Maybe 2 wrapper classes around strings would be more explicit.

**Review by BacLuc — APPROVED** [review]
When starting to write performance optimized code, there is always the tradeof between readability (here the inconsistency with the php way of writing code and that we use exceptions in the other places) and the gained performance.

A way for this would be:
1. Benchmark this endpoint (measurement done)
2. Show, that this is the crucial part were the performance can be gained, e.g. with a profiler. (Maybe there is a simpler place where we can gain more)
3. optimize the performance here
4. Benchmark again and show that it improved
5. Write a test that enforces this maybe uncommon style
6. Look how it behaves in production.

But i would say this is ok here and we look how the request time behaves when we have it running with real users against the real database.

But maybe there is more to gain with other configurations of the frankenphp container, as can be seen here: 
- https://github.com/ecamp/ecamp3/issues/5692

**Comment by pmattmann** [comment]
> When starting to write performance optimized code, there is always the tradeof between readability (here the inconsistency with the php way of writing code and that we use exceptions in the other places) and the gained performance.
> 
> A way for this would be:
> 
> 1. Benchmark this endpoint (measurement done)
> 2. Show, that this is the crucial part were the performance can be gained, e.g. with a profiler. (Maybe there is a simpler place where we can gain more)
> 3. optimize the performance here
> 4. Benchmark again and show that it improved
> 5. Write a test that enforces this maybe uncommon style
> 6. Look how it behaves in production.
> 
> But i would say this is ok here and we look how the request time behaves when we have it running with real users against the real database.
> 
> But maybe there is more to gain with other configurations of the frankenphp container, as can be seen here:
> 
> * [Analyse changes in the performance between php-fpm setup and frankenphp in prod #5692](https://github.com/ecamp/ecamp3/issues/5692)


I completely agree with the procedure you described!
That's exactly what I did.

Here is the initial measurement (XDebug Profiler) of  `localhost:3000/api/activities.jsonhal`
image

`RelatedCollectionLinkNormalizer::getRelatedCollectionHref()` is clearly a time consuming operation.


Here are the results after implementing the cache:

image

image

File: api/src/Serializer/Normalizer/RelatedCollectionLinkNormalizer.php (hunk 3 above)
Comment (pmattmann) [reply api/src/Serializer/Normalizer/RelatedCollectionLinkNormalizer.php]: would you prefer the following:
```php
$ok = $this->getRelatedCollectionHref($data, $rel, $context, out $href);
```

File: api/src/Serializer/Normalizer/RelatedCollectionLinkNormalizer.php (hunk 4 above)
Comment (pmattmann) [reply api/src/Serializer/Normalizer/RelatedCollectionLinkNormalizer.php]: would you prefer the following:

```php
  $ok = $this->getRelatedCollectionHref($data, $rel, $context, out $href);
  if (!$ok) {
    continue;
  }
  $normalized_data['_links'][$rel] = ['href' => $href];
```

File: api/src/Serializer/Normalizer/RelatedCollectionLinkNormalizer.php (hunk 3 above)
Comment (BacLuc) [reply api/src/Serializer/Normalizer/RelatedCollectionLinkNormalizer.php]: oh no, am strictly against out parameters.
i was just surprised.
Exceptions ~= (on the same level) return codes > out (or worse: in/out) parameters.

In php its just not idiomatic... but the idiomatic exceptions are slow as shown here.

File: api/src/Serializer/Normalizer/RelatedCollectionLinkNormalizer.php (hunk 5 above)
Comment (pmattmann) [reply api/src/Serializer/Normalizer/RelatedCollectionLinkNormalizer.php]: no, they are not both strings; I have converted it to a union type. thx

File: api/src/Serializer/Normalizer/RelatedCollectionLinkNormalizer.php (hunk 3 above)
Comment (BacLuc) [reply api/src/Serializer/Normalizer/RelatedCollectionLinkNormalizer.php]: it is ok, don't change it back after https://github.com/ecamp/ecamp3/pull/6536/commits/70a7155e779576c6723a15eb34ccff92bfafb192

**Comment by pmattmann** [comment]
@BacLuc 
If you agree with my last adjustments, please merge. 
thx

**Review by BacLuc — APPROVED** [review]

**Review by BacLuc — APPROVED** [review]

**Pull request merged**